### PR TITLE
Review 2 [C-01] Price Guard

### PR DESCRIPTION
**Pull request by lindsgil**
## Description of the change

Update the purchase and purchaseTo methods of the MinterSetPriceERC20V5 and MinterSetPricePolyptychERC20V5 minters to take the price as a parameter. This price is compared to the current minting price, and if it is lower than the current minting price then the transaction will revert.

### CODE REVIEW

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol [hunk 1]
```diff
@@ -450,12 +448,14 @@ contract MinterSetPriceERC20V5 is
      * @param to Address to be the new token's owner.
      * @param projectId Project ID to mint a token on.
      * @param coreContract Core contract address for the given project.
+     * @param purchasePricePerToken Price of token sent by purchaser
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:451]:
```suggestion
    * @param purchasePricePerToken Maximum price of token being allowed by the purchaser, no decimal places
```

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol [hunk 2]
```diff
@@ -492,6 +493,12 @@ contract MinterSetPriceERC20V5 is
             coreContract: coreContract
         });
 
+        // validate that the price sent is greater than or equal to the price per token
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:496]:
```suggestion
        // validate that the specified maximum price is greater than or equal to the price per token
```

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol [hunk 3]
```diff
@@ -492,6 +493,12 @@ contract MinterSetPriceERC20V5 is
...
+        require(
+            purchasePricePerToken >= pricePerTokenInWei,
+            "Must send amount greater than or equal to price per token"
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:499]: Prefer to keep revert messages <= 31 characters (helps keep everything in single slot, reducing contract bytecode size)
```suggestion
            "Only max price gte token price"
```

**Review by ryley-o — COMMENTED** [review]
a few minor nits, as well as a note about updating our other ERC20 minter 🙏 thank you!

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol (hunk 3 above)
Comment (lindsgil) [reply packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:499]: Makes sense, TY!

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol [hunk 4]
```diff
@@ -47,11 +46,7 @@ import {ReentrancyGuard} from "@openzeppelin-4.5/contracts/security/ReentrancyGu
  * configured on this minter, but they will not be able to mint tokens due to
  * checks performed by this minter's Minter Filter.
  */
-contract MinterSetPriceERC20V5 is
```
[the remark this answers is not in the document]
Comment (lindsgil) [reply packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:50]: Thank you for flagging! Updated this PR to include those changes 🙏

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol [hunk 5]
```diff
@@ -189,16 +184,19 @@ contract MinterSetPriceERC20V5 is
...
      * @param projectId Project ID to mint a token on.
      * @param coreContract Core contract address for the given project.
+     * @param purchasePricePerToken Maximum price of token being allowed by the purchaser, no decimal places
      * @return tokenId Token ID of minted token
      */
     function purchase(
         uint256 projectId,
-        address coreContract
+        address coreContract,
+        uint256 purchasePricePerToken
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol:193]: nit: super minor nit, but WDYT about calling this `maxPricePerToken` to be a little more idiomatic?
(applies to all purchase & purchaseTo functions across both minters)

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol [hunk 6]
```diff
@@ -8,7 +8,6 @@ pragma solidity 0.8.19;
 import {IGenArt721CoreContractV3_Base} from "../../interfaces/v0.8.x/IGenArt721CoreContractV3_Base.sol";
 import {IDelegationRegistry} from "../../interfaces/v0.8.x/IDelegationRegistry.sol";
 import {ISharedMinterV0} from "../../interfaces/v0.8.x/ISharedMinterV0.sol";
-import {ISharedMinterHolderV0} from "../../interfaces/v0.8.x/ISharedMinterHolderV0.sol";
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol:11]: nit: let's create an `ISharedMinterHolderERC20V0`, and import/conform to that 👍

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol [hunk 7]
```diff
@@ -398,21 +393,24 @@ contract MinterSetPricePolyptychERC20V5 is
...
+     * @param purchasePricePerToken Maximum price of token being allowed by the purchaser, no decimal places
      * @return tokenId Token ID of minted token
      */
     function purchase(
         uint256 projectId,
         address coreContract,
         address ownedNFTAddress,
-        uint256 ownedNFTTokenId
+        uint256 ownedNFTTokenId,
+        uint256 purchasePricePerToken
```
Comment (ryley-o) [inline packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol:404]: nit: very minor nit, but I think a little more straightforward to include the `maxPricePerToken` immediately after the `coreContract` arg here. That allows the ordering of args to be `[Fixed Price Minter Args],[Holder Args`].

**Review by ryley-o — APPROVED** [review]
LGTM - very minor nits are all that are left. Great work on this!

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol (hunk 5 above)
Comment (lindsgil) [reply packages/contracts/contracts/minter-suite/Minters/MinterSetPriceERC20V5.sol]: Good call! Updated

File: packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol (hunk 7 above)
Comment (lindsgil) [reply packages/contracts/contracts/minter-suite/Minters/MinterSetPricePolyptychERC20V5.sol]: Good call! Updated 🙏

**Pull request merged**